Sources/Algorithms/Stride.swift

```diff
@@ -0,0 +1,160 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// striding(by:)
+//===----------------------------------------------------------------------===//
+
+extension Collection {
+  /// Returns a collection stepping through the elements every `step` starting
+  /// at the first value. Any remainders of the stride will be trimmed.
+  ///
+  ///     (0...10).striding(by: 2) // == [0, 2, 4, 6, 8, 10]
+  ///     (0...10).striding(by: 3) // == [0, 3, 6, 9]
+  ///
+  /// - Complexity: O(1). Access to successive values is O(1) if the
+  /// collection conforms to `RandomAccessCollection`; otherwise,
+  /// O(_k_), where _k_ is the striding `step`.
+  ///
+  /// - Parameter step: The amount to step with each iteration.
+  /// - Returns: Returns a collection stepping through the elements by the
+  /// specified amount.
+  public func striding(by step: Int) -> Stride<Self> {
+    Stride(base: self, stride: step)
+  }
+}
+
+public struct Stride<Base: Collection> {
+  
+  public let base: Base
```

Comment:
To help simplify the issue, do we need to expose `base` at all?

Reply:
I was also unsure of the value of being able to refer to `base` in this instance, but since all of the other collection algorithms do I thought it would be worth keeping consistent. Was there a specific issue you think that it might simplify?

Reply:
Sorry, the specific issue is the question of equality, which @natecook1000 mentioned above.

Two values are equal when they behave the same way for all salient operations. For this purpose, we say that iteration order isn’t salient for sets but is salient for arrays, that the sign of zero isn’t salient for floating-point numbers, etc.

It’s a judgment call about what is salient here, and reasonable people can disagree whether differing `base` elements would matter; by contrast, if `base` isn’t recoverable, then there is no ambiguity at all.

Reply:
I like what you're saying here and I'd be up for dropping the access modifier from the property - I cannot fathom a situation where you would want to use it outside of the scope of the collection itself.

Reply:
Agreed on this — `base` can be internal, which removes any controversy here. (That should probably be the case for the rest of the collection wrappers as well.)

Reply:
Thanks both for your thoughts, [I've removed it](https://github.com/apple/swift-algorithms/pull/24/commits/dcdef508c0abc1839008ae14b7d6d6e960164735) from base here and we can follow up in another P/R for the rest of the collections

Reply:
>by contrast, if `base` isn’t recoverable, then there is no ambiguity at all.

I don't see how this matters in case the user knows what `base` is, even if `Stride` doesn't expose it. Couldn't there still be controversy about whether `(0...10).striding(by: 4)` should equal `(0...11).striding(by: 4)`?